File: app/src/main/java/com/zeyad/rxredux/screens/user/detail/UserDetailFragment.java
package com.zeyad.rxredux.screens.user.detail;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentActivity;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.transition.TransitionInflater;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.zeyad.gadapter.GenericRecyclerViewAdapter;
import com.zeyad.rxredux.R;
import com.zeyad.rxredux.core.redux.BaseEvent;
import com.zeyad.rxredux.core.redux.ErrorMessageFactory;
import com.zeyad.rxredux.screens.BaseFragment;
import com.zeyad.rxredux.screens.user.User;
import com.zeyad.rxredux.screens.user.ViewModelFactory;
import com.zeyad.rxredux.screens.user.list.UserListActivity;
import com.zeyad.rxredux.utils.Utils;
import com.zeyad.usecases.api.DataServiceFactory;

import org.parceler.Parcels;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;

import static com.zeyad.rxredux.core.redux.prelollipop.BaseActivity.UI_MODEL;

/**
 * A fragment representing a single Repository detail screen. This fragment is either contained in a
 * {@link UserListActivity} in two-pane mode (on tablets) or a {@link UserDetailActivity} on
 * handsets.
 */
public class UserDetailFragment extends BaseFragment<UserDetailState, UserDetailVM> {
    @BindView(R.id.linear_layout_loader)
    LinearLayout loaderLayout;

    @BindView(R.id.recyclerView_repositories)
    RecyclerView recyclerViewRepositories;

    private GenericRecyclerViewAdapter repositoriesAdapter;

    private RequestListener<String, GlideDrawable> requestListener = new RequestListener<String,
            GlideDrawable>() {
        @Override
        public boolean onException(Exception e, String model, Target<GlideDrawable> target,
                                   boolean isFirstResource) {
            return glideRequestListenerCore();
        }

        @Override
        public boolean onResourceReady(GlideDrawable resource, String model, Target<GlideDrawable> target,
                                       boolean isFromMemoryCache, boolean isFirstResource) {
            return glideRequestListenerCore();
        }
    };

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the fragment (e.g. upon
     * screen orientation changes).
     */
    public UserDetailFragment() {
    }

    public static UserDetailFragment newInstance(UserDetailState userDetailState) {
        UserDetailFragment userDetailFragment = new UserDetailFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(UI_MODEL, Parcels.wrap(userDetailState));
        userDetailFragment.setArguments(bundle);
        return userDetailFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        postponeEnterTransition();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setSharedElementEnterTransition(
                    TransitionInflater.from(getContext()).inflateTransition(android.R.transition.move));
        }
        //        setSharedElementReturnTransition(null); // supply the correct element for return transition
    }

    @NonNull
    @Override
    public ErrorMessageFactory errorMessageFactory() {
        return Throwable::getLocalizedMessage;
    }

    @Override
    public void initialize() {
        viewModel = ViewModelProviders.of(this,
                new ViewModelFactory(DataServiceFactory.getInstance())).get(UserDetailVM.class);
    }

    @Override
    public Observable<BaseEvent> events() {
        return Observable.just(new GetReposEvent(viewState.getUser().getLogin()));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.user_detail, container, false);
        ButterKnife.bind(this, rootView);
        setupRecyclerView();
        return rootView;
    }

    private void setupRecyclerView() {
        recyclerViewRepositories.setLayoutManager(new LinearLayoutManager(getContext()));
        repositoriesAdapter = new GenericRecyclerViewAdapter(
                (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE),
                new ArrayList<>()) {
            @Override
            public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
                return new RepositoryViewHolder(getLayoutInflater().inflate(viewType, parent, false));
            }
        };
        recyclerViewRepositories.setAdapter(repositoriesAdapter);
    }

    @Override
    public void renderSuccessState(UserDetailState userDetailState) {
        repositoriesAdapter.animateTo(userDetailState.getRepos());
        User user = userDetailState.getUser();
        if (userDetailState.isTwoPane()) {
            Utils.runIfNotNull((UserListActivity) getActivity(), activity -> {
                Toolbar appBarLayout = activity.findViewById(R.id.toolbar);
                if (appBarLayout != null) {
                    appBarLayout.setTitle(user.getLogin());
                }
                if (Utils.isNotEmpty(user.getAvatarUrl())) {
                    Glide.with(getContext()).load(user.getAvatarUrl()).dontAnimate().listener(requestListener)
                            .into(activity.imageViewAvatar);
                }
            });
        } else {
            Utils.runIfNotNull((UserDetailActivity) getActivity(), activity -> {
                CollapsingToolbarLayout appBarLayout = activity.collapsingToolbarLayout;
                appBarLayout.setTitle(user.getLogin());
                if (Utils.isNotEmpty(user.getAvatarUrl())) {
                    Glide.with(getContext()).load(user.getAvatarUrl()).dontAnimate().listener(requestListener)
                            .into(activity.imageViewAvatar);
                }
            });
        }
        //        applyPalette();
    }

    boolean glideRequestListenerCore() {
        Utils.runIfNotNull(getActivity(), FragmentActivity::supportStartPostponedEnterTransition);
        return false;
    }

    @Override
    public void toggleViews(boolean toggle) {
        loaderLayout.bringToFront();
        loaderLayout.setVisibility(toggle ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showError(String message) {
        showErrorSnackBar(message, loaderLayout, Snackbar.LENGTH_LONG);
    }

    private void applyPalette() {
        if (Utils.hasM()) {
            UserDetailActivity activity = (UserDetailActivity) getActivity();
            BitmapDrawable drawable = (BitmapDrawable) activity.imageViewAvatar.getDrawable();
            Bitmap bitmap = drawable.getBitmap();
            Palette.from(bitmap).generate(palette -> activity.findViewById(R.id.coordinator_detail)
                    .setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
                        if (v.getHeight() == scrollX) {
                            activity.toolbar.setTitleTextColor(palette.getLightVibrantColor(Color.TRANSPARENT));
                            activity.toolbar.setBackground(
                                    new ColorDrawable(palette.getLightVibrantColor(Color.TRANSPARENT)));
                        } else if (scrollY == 0) {
                            activity.toolbar.setTitleTextColor(0);
                            activity.toolbar.setBackground(null);
                        }
                    }));
        }
    }
}

File: app/src/main/java/com/zeyad/rxredux/screens/user/detail/UserDetailFragment.kt
package com.zeyad.rxredux.screens.user.detail

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.graphics.Palette
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.transition.TransitionInflater
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.GlideDrawable
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.zeyad.gadapter.GenericRecyclerViewAdapter
import com.zeyad.gadapter.ItemInfo
import com.zeyad.rxredux.R
import com.zeyad.rxredux.core.BaseEvent
import com.zeyad.rxredux.core.view.ErrorMessageFactory
import com.zeyad.rxredux.screens.BaseFragment
import com.zeyad.rxredux.screens.user.detail.UserDetailActivity.Companion.UI_MODEL
import com.zeyad.rxredux.screens.user.list.UserListActivity
import com.zeyad.rxredux.screens.user.list.UserListActivity2
import com.zeyad.rxredux.utils.hasM
import io.reactivex.Observable
import kotlinx.android.synthetic.main.user_detail.*
import kotlinx.android.synthetic.main.view_progress.*
import org.koin.android.architecture.ext.getViewModel
import java.util.*

/**
 * A fragment representing a single Repository detail screen. This fragment is either contained in a
 * [UserListActivity] in two-pane mode (on tablets) or a [UserDetailActivity] on
 * handsets.
 */
class UserDetailFragment : BaseFragment<UserDetailState, UserDetailVM>() {

    private lateinit var repositoriesAdapter: GenericRecyclerViewAdapter

    private val requestListener = object : RequestListener<String, GlideDrawable> {
        override fun onException(e: Exception,
                                 model: String,
                                 target: Target<GlideDrawable>,
                                 isFirstResource: Boolean): Boolean =
                glideRequestListenerCore()

        override fun onResourceReady(resource: GlideDrawable,
                                     model: String,
                                     target: Target<GlideDrawable>,
                                     isFromMemoryCache: Boolean,
                                     isFirstResource: Boolean): Boolean =
                glideRequestListenerCore()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        postponeEnterTransition()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            sharedElementEnterTransition = TransitionInflater.from(context).inflateTransition(android.R.transition.move)
        }
        //        setSharedElementReturnTransition(null); // supply the correct element for return transition
    }

    override fun errorMessageFactory(): ErrorMessageFactory = { throwable, _ -> throwable.localizedMessage }

    override fun initialize() {
        viewModel = getViewModel()
    }

    override fun initialState(): UserDetailState = arguments?.getParcelable(UI_MODEL)!!

    override fun events(): Observable<BaseEvent<*>> = Observable.just(viewState?.user?.login?.let { GetReposEvent(it) })

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.user_detail, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        recyclerView_repositories.layoutManager = LinearLayoutManager(context)
        repositoriesAdapter = object : GenericRecyclerViewAdapter(
                requireContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater,
                ArrayList<ItemInfo>()) {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GenericViewHolder<*> {
                return RepositoryViewHolder(layoutInflater.inflate(viewType, parent, false))
            }
        }
        recyclerView_repositories.adapter = repositoriesAdapter
    }

    override fun renderSuccessState(successState: UserDetailState) {
        repositoriesAdapter.setDataList(successState.repos, null)
        val user = successState.user
        if (successState.isTwoPane) {
            (activity as UserListActivity2).let { activity ->
                val appBarLayout = activity.findViewById<Toolbar>(R.id.toolbar)
                if (appBarLayout != null) {
                    appBarLayout.title = user.login
                }
                if (user.avatarUrl.isNotBlank()) {
                    Glide.with(context).load(user.avatarUrl).dontAnimate().listener(requestListener)
                            .into(activity.getImageViewAvatar())
                }
            }
        } else {
            (activity as UserDetailActivity).let { activity ->
                val appBarLayout = activity.getCollapsingToolbarLayout()
                appBarLayout.title = user.login
                if (user.avatarUrl.isNotBlank()) {
                    Glide.with(context).load(user.avatarUrl).dontAnimate().listener(requestListener)
                            .into(activity.getImageViewAvatar())
                }
            }
        }
    }

    internal fun glideRequestListenerCore(): Boolean {
        requireActivity().supportStartPostponedEnterTransition()
        return false
    }

    override fun toggleViews(isLoading: Boolean, event: BaseEvent<*>) {
        linear_layout_loader.bringToFront()
        linear_layout_loader.visibility = if (isLoading) View.VISIBLE else View.GONE
    }

    override fun showError(errorMessage: String, event: BaseEvent<*>) {
        showErrorSnackBar(errorMessage, linear_layout_loader, Snackbar.LENGTH_LONG)
    }

    private fun applyPalette() {
        if (hasM()) {
            val activity = activity as UserDetailActivity?
            val drawable = activity!!.getImageViewAvatar().drawable as BitmapDrawable
            val bitmap = drawable.bitmap
            Palette.from(bitmap).generate { palette ->
                palette?.let {
                    activity.findViewById<View>(R.id.coordinator_detail)
                            .setOnScrollChangeListener { v, scrollX, scrollY, _, _ ->
                                if (v.height == scrollX) {
                                    activity.getToolbar().setTitleTextColor(it
                                            .getLightVibrantColor(Color.TRANSPARENT))
                                    activity.getToolbar().background = ColorDrawable(it
                                            .getLightVibrantColor(Color.TRANSPARENT))
                                } else if (scrollY == 0) {
                                    activity.getToolbar().setTitleTextColor(0)
                                    activity.getToolbar().background = null
                                }
                            }
                }
            }
        }
    }

    companion object {

        fun newInstance(userDetailState: UserDetailState): UserDetailFragment {
            val userDetailFragment = UserDetailFragment()
            val bundle = Bundle()
            bundle.putParcelable(UI_MODEL, userDetailState)
            userDetailFragment.arguments = bundle
            return userDetailFragment
        }
    }
}
